File: c/physics/nbody/bench.c
/**
 * @file  bench.c
 * @brief Benchmark one nbody step — N=1000 bodies, O(n²) pairwise gravity.
 *
 * Outputs JSON: {"language":"c","example":"nbody","ns_per_op":...,"iterations":...}
 */

#include <stdio.h>
#include <time.h>
#include "nbody.h"

#define WARMUP  20
#define ITERS  200

static double now_ns(void) {
    struct timespec t;
    clock_gettime(CLOCK_MONOTONIC, &t);
    return t.tv_sec * 1e9 + t.tv_nsec;
}

int main(void) {
    static Body bodies[N_BODIES];

    nbody_init(bodies, N_BODIES);

    for (int i = 0; i < WARMUP; i++)
        nbody_step(bodies, N_BODIES, 1e-3);

    nbody_init(bodies, N_BODIES);

    double t0 = now_ns();
    for (int i = 0; i < ITERS; i++)
        nbody_step(bodies, N_BODIES, 1e-3);
    double t1 = now_ns();

    volatile double sink = bodies[0].x;
    (void)sink;

    double ns = (t1 - t0) / ITERS;
    printf("{\"language\":\"c\",\"example\":\"nbody\",\"ns_per_op\":%.2f,\"iterations\":%d}\n",
           ns, ITERS);
    return 0;
}

File: rust/physics/nbody/src/bench.rs
//! Benchmark one nbody step — N=1000 bodies, O(n²) pairwise gravity.
//!
//! Outputs JSON: {"language":"rust","example":"nbody","ns_per_op":...,"iterations":...}

mod physics;
use physics::{nbody_init, nbody_step};
use std::hint::black_box;
use std::time::Instant;

const WARMUP: usize = 20;
const ITERS:  usize = 200;

fn main() {
    let mut bodies = nbody_init();

    for _ in 0..WARMUP {
        nbody_step(&mut bodies, 1e-3);
    }
    bodies = nbody_init();

    let start = Instant::now();
    for _ in 0..ITERS {
        nbody_step(black_box(&mut bodies), 1e-3);
    }
    let elapsed = start.elapsed();

    let ns = elapsed.as_nanos() as f64 / ITERS as f64;
    println!(
        r#"{{"language":"rust","example":"nbody","ns_per_op":{:.2},"iterations":{}}}"#,
        ns, ITERS
    );
}
